core/amber/src/main/python/core/util/customized_queue/linked_blocking_multi_queue.py

```diff
@@ -0,0 +1,436 @@
+
+    def __len__(self) -> int:
+        return self.size()
+
+    def is_empty(self, key: Optional[str] = None) -> bool:
+        """
+        Check if the queue is empty, or check a specific SubQueue if
+        key is provided. This action acquires NO locks.
+
+        :param key: optional identifier of a SubQueue.
+        :raises KeyError for non-existing keys.
+        :return: Boolean representing empty or not.
+        """
+        return self.size(key) == 0
+
+    def add_sub_queue(self, key: str, priority: int) -> Optional[SubQueue]:
+        """
+        Create a new SubQueue if absent, with the key and priority.
+
+        :param key: SubQueue identifier for future reference.
+        :param priority: int value of priority, the lower number means the higher
+        priority.
+        :return: returns None if the key is new, or returns the previous SubQueue
+        mapped by the key if key is repeated.
+        """
+        sub_queue = self.SubQueue(key)
+        self.take_lock.acquire()
+
+        try:
+            old_queue = self.sub_queues.get(key)
+            self.sub_queues[key] = sub_queue
+            if old_queue is None:
+                i = 0
+                added = False
+                for pg in self.priority_groups:
+                    if pg.priority == priority:
+                        pg.add_queue(sub_queue)
+                        added = True
+                        break
+                    elif pg.priority > priority:
+                        new_pg = LinkedBlockingMultiQueue.PriorityGroup(priority)
+                        new_pg.add_queue(sub_queue)
+                        self.priority_groups.append(new_pg)
+                        added = True
+                        break
+
+                    i += 1
+                if not added:
+                    new_pg = LinkedBlockingMultiQueue.PriorityGroup(priority)
+                    new_pg.add_queue(sub_queue)
+                    self.priority_groups.append(new_pg)
+
+            return old_queue
+        finally:
+            self.take_lock.release()
+
+    def remove_sub_queue(self, key: str) -> SubQueue:
+        self.take_lock.acquire()
+        try:
+            removed: Optional[LinkedBlockingMultiQueue.SubQueue] = self.sub_queues.get(
+                key
+            )
+            if removed is not None:
+                del self.sub_queues[key]
+                removed.priority_group.remove_queue(removed)
+                if len(removed.priority_group.queues) == 0:
+                    self.priority_groups.remove(removed.priority_group)
+            return removed
+        finally:
+            self.take_lock.release()
+
+    def _get_sub_queue(self, key: str) -> SubQueue:
+        """
+        Get the SubQueue specified by the key.
+
+        :param key: the identifier of a SubQueue.
+        :raises KeyError for non-existing keys.
+        :return: the SubQueue.
+        """
+        return self.sub_queues[key]
```

Comment:
just wondering why does this start with underscore, I saw in java this function is public

Reply:
That's a leftover change from the previous version, where I designed the Internal queue to inherit LBMQ, so I made some of the methods to be private to prevent user of InternalQueue from calling it. Now since we are no longer doing inheritance, I can revert it back to public. 

Changed to public method.